File: app/src/test/java/com/astraedus/nudge/domain/emergency/EmergencyPassTest.kt
package com.astraedus.nudge.domain.emergency

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Tests for the pure emergency-pass ledger logic. The pass gates a protection-WEAKENING escape hatch
 * with GLOBAL (one-per-24h-across-all-apps) semantics, so both the eligibility boundary
 * (exact-cooldown, never-used, cross-app lockout) and robustness of the persisted-string parser
 * (malformed input must never throw; a legacy per-app ledger must migrate) are load-bearing.
 */
class EmergencyPassTest {

    private val cooldown = EmergencyPass.LOCKOUT_MS
    private val t0 = 1_000_000_000_000L // arbitrary fixed "now" base

    // ── parse / serialize round-trip ──

    @Test
    fun `empty string parses to empty map`() {
        assertEquals(emptyMap<String, Long>(), EmergencyPass.parse(""))
        assertEquals(emptyMap<String, Long>(), EmergencyPass.parse("   "))
    }

    @Test
    fun `single entry round-trips`() {
        val usage = mapOf("com.foo.bar" to 123456789L)
        assertEquals(usage, EmergencyPass.parse(EmergencyPass.serialize(usage)))
    }

    @Test
    fun `global record round-trips through serialize`() {
        val recorded = EmergencyPass.recordGlobal(t0)
        assertEquals(mapOf(EmergencyPass.GLOBAL_KEY to t0), recorded)
        assertEquals(recorded, EmergencyPass.parse(EmergencyPass.serialize(recorded)))
    }

    @Test
    fun `serialize of empty map is empty string`() {
        assertEquals("", EmergencyPass.serialize(emptyMap()))
    }

    @Test
    fun `malformed entries are ignored and never throw`() {
        // Missing value, missing key, non-numeric value, negative value, stray separators, no '='.
        val raw = "com.a=100;=200;com.b=;com.c=abc;;com.d=-5;garbage;com.e=300"
        val parsed = EmergencyPass.parse(raw)
        assertEquals(mapOf("com.a" to 100L, "com.e" to 300L), parsed)
    }

    @Test
    fun `entirely garbage input yields empty map`() {
        assertEquals(emptyMap<String, Long>(), EmergencyPass.parse(";;;===;no-equals;"))
    }

    // ── globalLastUsed (migration-safe MAX) ──

    @Test
    fun `globalLastUsed is null when never used`() {
        assertNull(EmergencyPass.globalLastUsed(emptyMap()))
    }

    @Test
    fun `globalLastUsed is the single global entry`() {
        assertEquals(t0, EmergencyPass.globalLastUsed(mapOf(EmergencyPass.GLOBAL_KEY to t0)))
    }

    @Test
    fun `globalLastUsed takes the MAX across a legacy per-app ledger (migration)`() {
        // Old per-app ledger: the most recent app's timestamp becomes the global last-used.
        val legacy = mapOf(
            "com.instagram.android" to t0,
            "com.zhiliaoapp.musically" to t0 + 5_000,
            "com.google.android.youtube" to t0 - 10_000
        )
        assertEquals(t0 + 5_000, EmergencyPass.globalLastUsed(legacy))
    }

    // ── canUseGlobal ──

    @Test
    fun `canUseGlobal true when never used before`() {
        assertTrue(EmergencyPass.canUseGlobal(emptyMap(), t0, cooldown))
    }

    @Test
    fun `canUseGlobal false immediately after use`() {
        val usage = EmergencyPass.recordGlobal(t0)
        assertFalse(EmergencyPass.canUseGlobal(usage, t0, cooldown))
    }

    @Test
    fun `canUseGlobal false just before cooldown elapses`() {
        val usage = EmergencyPass.recordGlobal(t0)
        assertFalse(EmergencyPass.canUseGlobal(usage, t0 + cooldown - 1, cooldown))
    }

    @Test
    fun `canUseGlobal true exactly at cooldown boundary`() {
        val usage = EmergencyPass.recordGlobal(t0)
        assertTrue(EmergencyPass.canUseGlobal(usage, t0 + cooldown, cooldown))
    }

    @Test
    fun `canUseGlobal is GLOBAL - using on any app locks out all apps`() {
        // A per-app ledger with a single recent entry (e.g. pass used on Instagram) still blocks the
        // pass for every other app because the lockout is global, not per-package.
        val usedOnInstagram = mapOf("com.instagram.android" to t0)
        // No matter which app's block screen we are on, the pass is locked until cooldown elapses.
        assertFalse(EmergencyPass.canUseGlobal(usedOnInstagram, t0 + 1, cooldown))
        assertTrue(EmergencyPass.canUseGlobal(usedOnInstagram, t0 + cooldown, cooldown))
    }

    // ── nextAvailableGlobalMs ──

    @Test
    fun `nextAvailableGlobalMs is zero when never used`() {
        assertEquals(0L, EmergencyPass.nextAvailableGlobalMs(emptyMap(), t0, cooldown))
    }

    @Test
    fun `nextAvailableGlobalMs is full cooldown immediately after use`() {
        val usage = EmergencyPass.recordGlobal(t0)
        assertEquals(cooldown, EmergencyPass.nextAvailableGlobalMs(usage, t0, cooldown))
    }

    @Test
    fun `nextAvailableGlobalMs decreases as time passes`() {
        val usage = EmergencyPass.recordGlobal(t0)
        val elapsed = 3_600_000L // 1h
        assertEquals(cooldown - elapsed, EmergencyPass.nextAvailableGlobalMs(usage, t0 + elapsed, cooldown))
    }

    @Test
    fun `nextAvailableGlobalMs is zero once cooldown has passed`() {
        val usage = EmergencyPass.recordGlobal(t0)
        assertEquals(0L, EmergencyPass.nextAvailableGlobalMs(usage, t0 + cooldown, cooldown))
        assertEquals(0L, EmergencyPass.nextAvailableGlobalMs(usage, t0 + cooldown + 5_000, cooldown))
    }

    @Test
    fun `nextAvailableGlobalMs uses the MAX timestamp of a legacy ledger`() {
        val legacy = mapOf("com.a" to t0, "com.b" to t0 + 10_000)
        // Locked out relative to the most recent (com.b) entry.
        assertEquals(cooldown - 10_000, EmergencyPass.nextAvailableGlobalMs(legacy, t0 + 20_000, cooldown))
    }

    // ── recordGlobal ──

    @Test
    fun `recordGlobal collapses to a single global entry`() {
        val result = EmergencyPass.recordGlobal(t0)
        assertEquals(mapOf(EmergencyPass.GLOBAL_KEY to t0), result)
    }

    @Test
    fun `recordGlobal output round-trips through serialize and canUseGlobal`() {
        val recorded = EmergencyPass.recordGlobal(t0)
        val reparsed = EmergencyPass.parse(EmergencyPass.serialize(recorded))
        assertFalse(EmergencyPass.canUseGlobal(reparsed, t0, cooldown))
        assertTrue(EmergencyPass.canUseGlobal(reparsed, t0 + cooldown, cooldown))
    }

    // ── constants sanity ──

    @Test
    fun `constants match the spec`() {
        assertEquals(120_000L, EmergencyPass.PASS_DURATION_MS)
        assertEquals(86_400_000L, EmergencyPass.LOCKOUT_MS)
        assertEquals("*", EmergencyPass.GLOBAL_KEY)
    }
}

File: extension/tests/core/emergencyPass.test.ts
import { describe, expect, it } from 'vitest';
import {
  GLOBAL_KEY,
  LOCKOUT_MS,
  PASS_DURATION_MS,
  canUseGlobal,
  globalLastUsed,
  nextAvailableGlobalMs,
  parse,
  recordGlobal,
  serialize,
} from '../../src/core/emergencyPass';

// Tests for the pure emergency-pass ledger logic. The pass gates a protection-WEAKENING escape
// hatch with GLOBAL (one-per-24h-across-all-sites) semantics, so both the eligibility boundary
// (exact-cooldown, never-used, cross-site lockout) and robustness of the persisted-ledger parser
// (malformed input must never throw; a legacy per-site ledger must migrate) are load-bearing.

const cooldown = LOCKOUT_MS;
const t0 = 1_000_000_000_000; // arbitrary fixed "now" base

describe('emergencyPass: parse() / serialize() round-trip', () => {
  it('empty string parses to an empty object', () => {
    expect(parse('')).toEqual({});
    expect(parse('   ')).toEqual({});
  });

  it('single entry round-trips', () => {
    const usage = { 'foo.com': 123456789 };
    expect(parse(serialize(usage))).toEqual(usage);
  });

  it('global record round-trips through serialize', () => {
    const recorded = recordGlobal(t0);
    expect(recorded).toEqual({ [GLOBAL_KEY]: t0 });
    expect(parse(serialize(recorded))).toEqual(recorded);
  });

  it('serialize of an empty object is an empty string', () => {
    expect(serialize({})).toBe('');
  });

  it('malformed entries are ignored and never throw', () => {
    // Missing value, missing key, non-numeric value, negative value, stray separators, no '='.
    const raw = 'a.com=100;=200;b.com=;c.com=abc;;d.com=-5;garbage;e.com=300';
    expect(() => parse(raw)).not.toThrow();
    const parsed = parse(raw);
    expect(parsed).toEqual({ 'a.com': 100, 'e.com': 300 });
  });

  it('entirely garbage input yields an empty object and never throws', () => {
    expect(() => parse(';;;===;no-equals;')).not.toThrow();
    expect(parse(';;;===;no-equals;')).toEqual({});
  });
});

describe('emergencyPass: globalLastUsed() (migration-safe MAX)', () => {
  it('is null when never used', () => {
    expect(globalLastUsed({})).toBeNull();
  });

  it('is the single global entry', () => {
    expect(globalLastUsed({ [GLOBAL_KEY]: t0 })).toBe(t0);
  });

  it('takes the MAX across a legacy per-site ledger (migration)', () => {
    // Old per-site ledger: the most recent site's timestamp becomes the global last-used.
    const legacy = {
      'instagram.com': t0,
      'tiktok.com': t0 + 5_000,
      'youtube.com': t0 - 10_000,
    };
    expect(globalLastUsed(legacy)).toBe(t0 + 5_000);
  });
});

describe('emergencyPass: canUseGlobal()', () => {
  it('true when never used before', () => {
    expect(canUseGlobal({}, t0, cooldown)).toBe(true);
  });

  it('false immediately after use', () => {
    const usage = recordGlobal(t0);
    expect(canUseGlobal(usage, t0, cooldown)).toBe(false);
  });

  it('false just before cooldown elapses', () => {
    const usage = recordGlobal(t0);
    expect(canUseGlobal(usage, t0 + cooldown - 1, cooldown)).toBe(false);
  });

  it('true exactly at the cooldown boundary', () => {
    const usage = recordGlobal(t0);
    expect(canUseGlobal(usage, t0 + cooldown, cooldown)).toBe(true);
  });

  it('is GLOBAL — using on any site locks out all sites', () => {
    // A per-site ledger with a single recent entry (e.g. pass used on Instagram) still blocks
    // the pass for every other site because the lockout is global, not per-domain.
    const usedOnInstagram = { 'instagram.com': t0 };
    // No matter which site's block screen we are on, the pass is locked until cooldown elapses.
    expect(canUseGlobal(usedOnInstagram, t0 + 1, cooldown)).toBe(false);
    expect(canUseGlobal(usedOnInstagram, t0 + cooldown, cooldown)).toBe(true);
  });
});

describe('emergencyPass: nextAvailableGlobalMs()', () => {
  it('is zero when never used', () => {
    expect(nextAvailableGlobalMs({}, t0, cooldown)).toBe(0);
  });

  it('is the full cooldown immediately after use', () => {
    const usage = recordGlobal(t0);
    expect(nextAvailableGlobalMs(usage, t0, cooldown)).toBe(cooldown);
  });

  it('decreases as time passes (decay toward zero)', () => {
    const usage = recordGlobal(t0);
    const elapsed = 3_600_000; // 1h
    expect(nextAvailableGlobalMs(usage, t0 + elapsed, cooldown)).toBe(cooldown - elapsed);
  });

  it('is zero once the cooldown has passed', () => {
    const usage = recordGlobal(t0);
    expect(nextAvailableGlobalMs(usage, t0 + cooldown, cooldown)).toBe(0);
    expect(nextAvailableGlobalMs(usage, t0 + cooldown + 5_000, cooldown)).toBe(0);
  });

  it('uses the MAX timestamp of a legacy ledger', () => {
    const legacy = { 'a.com': t0, 'b.com': t0 + 10_000 };
    // Locked out relative to the most recent (b.com) entry.
    expect(nextAvailableGlobalMs(legacy, t0 + 20_000, cooldown)).toBe(cooldown - 10_000);
  });
});

describe('emergencyPass: recordGlobal()', () => {
  it('collapses to a single global entry', () => {
    expect(recordGlobal(t0)).toEqual({ [GLOBAL_KEY]: t0 });
  });

  it('overwrites any prior per-site entries (collapse, not merge)', () => {
    // recordGlobal is a pure constructor — the "collapse" behavior is verified by the caller
    // (settings write) discarding the old ledger entirely and using this return value as-is.
    const fresh = recordGlobal(t0);
    expect(Object.keys(fresh)).toEqual([GLOBAL_KEY]);
  });

  it('output round-trips through serialize and canUseGlobal', () => {
    const recorded = recordGlobal(t0);
    const reparsed = parse(serialize(recorded));
    expect(canUseGlobal(reparsed, t0, cooldown)).toBe(false);
    expect(canUseGlobal(reparsed, t0 + cooldown, cooldown)).toBe(true);
  });
});

describe('emergencyPass: constants', () => {
  it('match the spec', () => {
    expect(PASS_DURATION_MS).toBe(120_000);
    expect(LOCKOUT_MS).toBe(86_400_000);
    expect(GLOBAL_KEY).toBe('*');
  });
});
